useDragHooks/useDropHooks docs (#3012)
adobe/react-spectrum: packages/@react-spectrum/dnd/docs/dragAndDrop.mdx

> {/* Copyright 2022 Adobe. All rights reserved.

Reviewer: I think we could just call this `dnd.mdx`, like we have `ssr.mdx`

Author: Sounds good, I'll do this last so the comments don't get outdated

export default Layout;

import ChevronRight from '@spectrum-icons/workflow/ChevronRight';
import docs from 'docs:@react-spectrum/dnd';
import dndDocs from 'docs:@react-types/shared/src/dnd.d.ts';
import {FunctionAPI, PageDescription, TypeContext, InterfaceType} from '@react-spectrum/docs';
import {Keyboard} from '@react-spectrum/text';

---
category: Concepts
keywords: [drag and drop, dnd]
after_version: 3.0.0
---

# Drag and Drop

This page describes how to enable drag and drop functionality for the various React Spectrum components that support it.

## Introduction

Drag and drop is an operation performed when a user "grabs" a UI element via their pointing device and drops it on a different location or element.
It has become a common interaction in many modern applications due to its flexibility across a variety of use cases but can come saddled with
many accessibility issues, including but not limited to keyboard interaction parity and difficulties mimicking drag operations with mouse emulators.
React Spectrum components that support drag and drop operations seek to address these issues by providing full keyboard and screen reader drag and drop
support.

> If you are interested in adding drag and drop support to your own components, feel free to check out the [drag and drop react-aria hooks](https://www.npmjs.com/package/@react-aria/dnd)!

Reviewer: I'd omit this for now until we have docs.

## API

<FunctionAPI function={docs.exports.useDragHooks} links={docs.links} />
<FunctionAPI function={docs.exports.useDropHooks} links={docs.links} />

## Usage

Enabling drag and drop for a supported React Spectrum component differs slightly from the typical event handler prop pattern that you may be familiar with.
Instead of providing each event handler directly to the component, you must first import `useDragHooks` and `useDropHooks` from `@react-spectrum/dnd` package
and provide those hooks with your desired options.

`useDragHooks` supports the following options:
<TypeContext.Provider value={dndDocs.links}>
  <InterfaceType properties={dndDocs.exports.DraggableCollectionProps.properties} />
</TypeContext.Provider>

`useDropHooks` supports the following options:
<TypeContext.Provider value={dndDocs.links}>
  <InterfaceType properties={dndDocs.exports.DroppableCollectionProps.properties} />

> </TypeContext.Provider>

Author: Open question: do we feel like this is sufficient information for someone to successfully use `useDragHooks` and `useDropHooks`? Do we need more examples to demonstrate each option available to these hooks/ explanations of what each option is used for?

I'm personally a bit concerned about `getItems`, `getAllowedDropOperations`, and `getDropOperation`

Reviewer 2: I think a little blurb about each one with a real life application, not necessarily a full blown example, would be useful

that removes the dragged items from the list only upon a successful drop operation.

<details>
  <summary style={{fontWeight: 'bold'}}><ChevronRight size="S" /> Show code</summary>

```tsx example export=true render=false
import Folder from '@spectrum-icons/workflow/Folder';
import {Item, ListView} from '@react-spectrum/list';
import {Text} from '@adobe/react-spectrum';
import {useDragHooks} from '@react-spectrum/dnd';
import {useListData} from '@react-stately/data';

function DraggableList(props) {
  let {
    items,
    ...otherProps
  } = props;
  let list = useListData({
    initialItems: items || [
      {id: 'a', textValue: 'Adobe Photoshop', type: 'file'},
      {id: 'b', textValue: 'Adobe XD', type: 'file'},
      {id: 'c', textValue: 'Documents', type: 'folder'},
      {id: 'd', textValue: 'Adobe InDesign', type: 'file'},
      {id: 'e', textValue: 'Utilities', type: 'folder'}
    ]
  });

  let getItems = (keys) => [...keys].map(key => {
    let item = list.getItem(key);
    return {
      'text/plain': item.textValue,
      [item.type]: item.id
    };

>   });

Author: perhaps I'm misusing this here? Just trying to imagine the best way to communicate an item's non-text information since the stuff returned here shows up in the item's types in onDrop and then accessed via getText. Kinda feels weird passing the `id` like this

Perhaps I am misrepresenting the items of the ListView? Looking at https://github.com/adobe/react-spectrum/blob/b4977a1d6279337d75d6b8b5c2645ffbb88fdee8/packages/%40react-types/shared/src/dnd.d.ts#L56-L74

Reviewer: JSON.stringify it

    }
  };

  let getDropOperation = (target, types) => {
    if (types.has('folder')) {
      return 'cancel';
    }

    return 'move';
  };

  let dropHooks = useDropHooks({
    getDropOperation,
    onDrop
  });

  return (
    <ListView
      aria-label="Droppable list view example"
      width="size-3600"
      height="size-3600"
      selectionMode="multiple"
      items={list.items}
      dropHooks={dropHooks}
      {...otherProps}>
      {item => (
        <Item key={item.key} textValue={item.textValue}>
          {item.type === 'folder' && <Folder />}
          <Text>{item.textValue}</Text>
        </Item>
      )}
    </ListView>
  );
}
```
</details>

To perform a drag and drop operations via a keyboard, first select the items to be dragged. You can then start the drag operation by focusing the drag handle on any of the
selected rows and hitting <Keyboard>Enter</Keyboard>/<Keyboard>Space</Keyboard>. Once a drag operation is started, <Keyboard>Tab</Keyboard> can be used to cycle through valid
drop targets. Hitting <Keyboard>Enter</Keyboard> will then confirm the drop operation on the focused drop target. For screen readers, please follow the instructions announced

> when focusing the row's drag handle to begin a drag operation. Go ahead and try out drag and drop in the example below!

Author: Including this description of how to drag and drop with the other interaction modes since it feels like it would be pretty common across our drag and drop compatible components

Reviewer: If we put this at the top of the page, it probably won't need to be repeated here.

Author: Ok I'll put it in the intro section

{/* Copyright 2022 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License. */}

import {Layout} from '@react-spectrum/docs';
export default Layout;

import ChevronRight from '@spectrum-icons/workflow/ChevronRight';
import docs from 'docs:@react-spectrum/dnd';
import dndDocs from 'docs:@react-types/shared/src/dnd.d.ts';
import {FunctionAPI, PageDescription, TypeContext, InterfaceType} from '@react-spectrum/docs';
import {Keyboard} from '@react-spectrum/text';

---
category: Concepts
keywords: [drag and drop, dnd]
after_version: 3.0.0

> ---

Reviewer: I think we should set the `order` property here so Drag and Drop is not the first concept listed in the sidebar. I'd probably put it last.

# Drag and Drop

This page describes how to enable drag and drop functionality for the various React Spectrum components that support it.

## Introduction

Drag and drop is an operation performed when a user "grabs" a UI element via their pointing device and drops it on a different location or element.
It has become a common interaction in many modern applications due to its flexibility across a variety of use cases but can come saddled with
many accessibility issues, including but not limited to keyboard interaction parity and difficulties mimicking drag operations with mouse emulators.
React Spectrum components that support drag and drop operations seek to address these issues by providing full keyboard and screen reader drag and drop
support.

> If you are interested in adding drag and drop support to your own components, check out the [drag and drop react-aria hooks](https://www.npmjs.com/package/@react-aria/dnd)!

Reviewer: I'd omit this until we have docs.

> 

Reviewer: I think we need a section to explain the overall concepts of drag and drop:

* drag sources and drop targets
* drag types: standard mime types, custom application-specific types, providing data in multiple formats 
* dragging multiple items at once
* drag previews
* drop operations: move, copy, link, cancel
* drop positions: on, between, root
* overall, at a very brief high level, how these all work across traditional dnd surfaces, keyboards, and screen readers

Perhaps borrow some content from the [RFC](https://github.com/adobe/react-spectrum/blob/main/rfcs/2020-v3-dnd.md)

Author: sounds good, I'll do this in a followup PR

## API

<FunctionAPI function={docs.exports.useDragHooks} links={docs.links} />
<FunctionAPI function={docs.exports.useDropHooks} links={docs.links} />

> ## Usage

Reviewer: I think the API and Usage sections here could be combined, with the function API right above the props table for each.

Author: Just wanna note that the current structure mimics the aria hook docs structure. I'm fine with combining the sections (I agree that it would be clearer), but lemme know if we should remain consistent w/ the hooks doc structure instead

Reviewer: I feel this is different since this page is more conceptual

Enabling drag and drop for a supported React Spectrum component differs slightly from the typical event handler prop pattern that you may be familiar with.
Instead of providing each event handler directly to the component, you must first import `useDragHooks` and `useDropHooks` from `@react-spectrum/dnd` package
and provide those hooks with your desired options.

`useDragHooks` supports the following options:
<TypeContext.Provider value={dndDocs.links}>
  <InterfaceType properties={dndDocs.exports.DraggableCollectionProps.properties} />
</TypeContext.Provider>

`useDropHooks` supports the following options:
<TypeContext.Provider value={dndDocs.links}>
  <InterfaceType properties={dndDocs.exports.DroppableCollectionProps.properties} />
</TypeContext.Provider>

`useDragHooks` will then provide you with a set of drag hooks that you can pass to the component via its `dragHooks` prop, thus enabling drag operations for the component.
Drop operations can be enabled in the same way by passing the hooks from `useDropHooks` to the component's `dropHooks` prop.

> Of the various hook options above, `getAllowedDropOperations` and `getDropOperation` maybe of particular interest since it allows you to specify what kinds of drag and drop operations you want to allow.

Reviewer 2: ```suggestion
Of the various hook options above, `getAllowedDropOperations` and `getDropOperation` may be of particular interest since it allows you to specify what kinds of drag and drop operations you want to allow.
```

For instance, you may have a draggable collection of items that allows `move` and `copy` operations but you need a way to know whether or not you should be removing the dragged items
from the list after a drop operation. A drop target that only allows `copy` operations, such as a file upload drop zone, would be able to return `copy` from its `getDropOperation` and communicate that
to your draggable collection's `onDragEnd` handler, letting it know that it shouldn't remove the dragged items from the list. Alternatively, a drop target that allows `move` operations, like in the
[example](#example) below, would return `move` from its `getDropOperation` and thus inform your draggable collection to remove the dragged items from it list.

You may be wondering why we didn't simply import the drag and drop hooks within the components themselves. The reason behind this is to reduce the bundle size of
the component for users who don't need the drag and drop support for their component.

The following list shows which components currently support drag and drop:
-  [ListView](ListView.html#drag-and-drop)

> ## Example

Reviewer: I'd move the example above the API/Usage section similar to how we put the prop table at the end in most spectrum docs.

Author: See https://github.com/adobe/react-spectrum/pull/3012#discussion_r879964594, I'm fine with mimicking the spectrum component docs instead as well

The example below illustrates how one would use `useDragHooks` and `useDropHooks` to enable dragging files from one ListView to another. For the first ListView, we want to make the rows draggable and
have the dragged rows removed from the list upon a successful drop. To accomplish this, we first want to specify the types of data associated with each item via the array returned by `getItems`. In this case,
we add a `text/plain` type that matches the text value associated with each item. The item's type is also included with the item's `id` mapped to it. We also provide a `onDragEnd` event handler to `useDragHooks`
that removes the dragged items from the list only upon a successful drop operation.

> <details>

Reviewer: Why are the codes hidden?

Author: Felt like they took up a lot of space visually between the explanation and the example itself, but will remove the <details> now that we are moving the example to the top

  <summary style={{fontWeight: 'bold'}}><ChevronRight size="S" /> Show code</summary>

```tsx example export=true render=false
import Folder from '@spectrum-icons/workflow/Folder';
import {Item, ListView} from '@react-spectrum/list';
import {Text} from '@adobe/react-spectrum';
import {useDragHooks} from '@react-spectrum/dnd';
import {useListData} from '@react-stately/data';

function DraggableList(props) {
  let {
    items,
    ...otherProps
  } = props;
  let list = useListData({
    initialItems: items || [
      {id: 'a', textValue: 'Adobe Photoshop', type: 'file'},
      {id: 'b', textValue: 'Adobe XD', type: 'file'},
      {id: 'c', textValue: 'Documents', type: 'folder'},
      {id: 'd', textValue: 'Adobe InDesign', type: 'file'},
      {id: 'e', textValue: 'Utilities', type: 'folder'}
    ]
  });

  let getItems = (keys) => [...keys].map(key => {
    let item = list.getItem(key);
    return {
      'text/plain': item.textValue,
      [item.type]: item.id
    };
  });

  let onDragEnd = (e) => {
    if (e.dropOperation === 'move') {
      list.remove(...e.keys);
    }
  };
  let dragHooks = useDragHooks({
    getItems,

>     onDragEnd

Reviewer: I think it might be easier to read if the event handlers/functions are inline here.

      aria-label="Droppable list view example"
      width="size-3600"
      height="size-3600"
      selectionMode="multiple"
      items={list.items}
      dropHooks={dropHooks}
      {...otherProps}>
      {item => (
        <Item key={item.key} textValue={item.textValue}>
          {item.type === 'folder' && <Folder />}
          <Text>{item.textValue}</Text>
        </Item>
      )}
    </ListView>
  );
}
```
</details>

To perform a drag and drop operations via a keyboard, first select the items to be dragged. You can then start the drag operation by focusing the drag handle on any of the
selected rows and hitting <Keyboard>Enter</Keyboard>/<Keyboard>Space</Keyboard>. Once a drag operation is started, <Keyboard>Tab</Keyboard> can be used to cycle through valid
drop targets. Hitting <Keyboard>Enter</Keyboard> will then confirm the drop operation on the focused drop target. For screen readers, please follow the instructions announced
when focusing the row's drag handle to begin a drag operation. Go ahead and try out drag and drop in the example below!

```tsx example
import {Flex} from '@react-spectrum/layout';

<Flex gap="size-300">
  <DraggableList />
  <DroppableList />

> </Flex>

Reviewer: We could consider putting the rendered example first if the description code is long. Then people can try it out first and then read how it works after?

{/* Copyright 2022 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License. */}

import {Layout} from '@react-spectrum/docs';
export default Layout;

import ChevronRight from '@spectrum-icons/workflow/ChevronRight';
import docs from 'docs:@react-spectrum/dnd';
import dndDocs from 'docs:@react-types/shared/src/dnd.d.ts';
import {FunctionAPI, PageDescription, TypeContext, InterfaceType} from '@react-spectrum/docs';
import {Keyboard} from '@react-spectrum/text';

---
category: Concepts
keywords: [drag and drop, dnd]
after_version: 3.0.0
order: 6
---

> # Drag and Drop

Author: Open to opinions on whether or not enough information is covered. My goal was to make it introductory level and not to get too deep into the details since that might put readers off

This page describes how to enable drag and drop functionality for the various React Spectrum components that support it.

## Introduction

> Drag and drop is an operation performed when a user "grabs" a UI element via their pointing device and drops it on a different location or element.

Reviewer 2: pointing device sounds funny, maybe we can just drop it for the list? otherwise, telling people what a pointing device is

```suggestion
Drag and drop is an operation performed when a user "grabs" a UI element via their pointing device, mouse, finger, keyboard, etc, and drops it on a different location or element.
```

Author: Maybe I'll just say "when a user "grabs" a UI element and drops it"

Due to its flexibility across a variety of use cases, it has become a common interaction in many modern applications but can come saddled with
many accessibility issues, including but not limited to keyboard interaction parity and difficulties mimicking drag operations with mouse emulators.
React Spectrum collection components that support drag and drop operations seek to address these issues by providing full keyboard and screen reader drag and drop
support.

## Drag and Drop Concepts

Before we dive into how to enable drag and drop in React Spectrum, let's touch briefly on the terminology and concepts of drag and drop. In a drag and drop operation, there is

> a **drag source** and a **drop target**. The drag source is the starting location of your dragged data and the drop target is its intended destination. The dragged data

Reviewer 2: I don't think location makes much sense here?
```suggestion
a **drag source** and a **drop target**. The drag source is the data associated with the element where the drag was started and the drop target is its intended destination. The dragged data
```

Author: Not quite, in this case the ListView is a drag source, with the items inside being valid draggable items

      {...otherProps}>
      {item => (
        <Item key={item.key}>
          {item.textValue}
        </Item>
      )}
    </ListView>
  );
}
```

### Creating the droppable list

For the second ListView, we want to make it droppable but have it only accept root level drops. Similar to the draggable ListView, we'll start by initializing the default item list
via [useListData](/react-stately/useListData.html).

```tsx example render=false
let list = useListData({
  initialItems: items || [
    {id: 'f', textValue: 'Adobe AfterEffects', type: 'file'},
    {id: 'g', textValue: 'Adobe Illustrator', type: 'file'},
    {id: 'h', textValue: 'Adobe Lightroom', type: 'file'},
    {id: 'i', textValue: 'Adobe Premiere Pro', type: 'file'},
    {id: 'j', textValue: 'Adobe Fresco', type: 'file'}
  ]
});
```

To implement root level only drops, we create a `getDropOperation` function that returns `"cancel"` for any drop targets other than `root` or if the dragged item types doesn't include
`'file'`.

```tsx example render=false
function getDropOperation(target, types) {
  if (target.type !== 'root' || !types.has('file')) {
    return 'cancel';
  }

  return 'move';
}
```

> Next, we create a `onDrop` event handler to process the successfully dropped items. This `onDrop` handler first checks each dropped item's kind and type, extracting the item's

Reviewer 2: ```suggestion
Next, we create an `onDrop` event handler to process the successfully dropped items. This `onDrop` handler first checks each dropped item's kind and type, extracting the item's
```

drop targets. Hitting <Keyboard>Enter</Keyboard> will then confirm the drop operation on the focused drop target. For screen readers, please follow the instructions announced
when focusing the row's drag handle to begin a drag operation. Go ahead and try out drag and drop in the example below!

## Drag and Drop Concepts

> TODO: add stuff from RFC

Reviewer 2: ?

Author: That looks like an outdated diff, try refreshing. I'm also doing a final edit sweep, so probably hold off till tmrw haha
